src/IO/LasLoader/lasloader.cpp

```diff
@@ -0,0 +1,244 @@
+    auto geometry =
+        make_unique<GeometryData>( "LAS_PC_" + to_string( ++nameId ), GeometryData::POINT_CLOUD );
+    if ( geometry == nullptr ) {
+        LOG( logERROR ) << "could not create geometry";
+        return nullptr;
+    }
+    geometry->setFrame( Ra::Core::Transform::Identity() );
+
+    ifstream stream( filename, ios::out | ios::binary );
+
+    if ( !stream.is_open() ) {
+        throw runtime_error( "Could not open binary ifstream to path " + filename );
+        return nullptr;
+    }
+    else {
+        LOG( logINFO ) << "---Beginning file loading---";
+    }
+
+    /*header part*/
+    char buffer[48];
+
+    // reading minor version (needed to check data format)
+    stream.seekg( 0 );
+    stream.seekg( 25 );
+    stream.read( buffer, 1 );
+    unsigned char minor = *(unsigned char*)buffer;
+
+    if ( ( minor < 1 ) || ( minor > 4 ) ) {
+        delete fileData;
+        LOG( logERROR ) << "LAS version 1." << (int)minor << " unsupported";
+        return nullptr;
+    }
+    else {
+        LOG( logINFO ) << "LAS version 1." << (int)minor;
+    }
+
+    // reading distance from file start to first point record
+    stream.seekg( 0 );
+    stream.seekg( 96 );
+    stream.read( buffer, 4 );
+    unsigned int offset = *(unsigned int*)buffer;
+
+    // reading point data format
+    stream.seekg( 0 );
+    stream.seekg( 104 );
+    stream.read( buffer, 1 );
+    unsigned char data_format = *(unsigned char*)buffer;
+
+    if ( ( data_format < 0 ) || ( data_format > 3 && minor < 3 ) ||
+         ( data_format > 5 && minor == 3 ) || ( data_format > 6 && minor == 4 ) ) {
+        delete fileData;
+        throw runtime_error( "Corrupted file. Unvalid data format" );
+        return nullptr;
+    }
+    else {
+        LOG( logINFO ) << "Data format " << (int)data_format;
+        LOG( logINFO ) << "Loading properties \"x\", \"y\", \"z\",";
+    }
+
+    // reading data record length (bytes)
+    stream.read( buffer, 2 );
+    unsigned short data_len = *(unsigned short*)buffer;
+
+    // reading total number of data records
+    stream.read( buffer, 4 );
+    unsigned int nb_data = *(unsigned int*)buffer;
+
+    // reading scales and offsets (used for computing coordinates)
+    stream.seekg( 0 );
+    stream.seekg( 131 );
+    stream.read( buffer, 48 );
+
+    double scale_x  = *(double*)buffer;
+    double scale_y  = *(double*)( buffer + 8 );
+    double scale_z  = *(double*)( buffer + 16 );
+    double offset_x = *(double*)( buffer + 24 );
+    double offset_y = *(double*)( buffer + 32 );
+    double offset_z = *(double*)( buffer + 40 );
+
+    /*loading properties*/
+    char* point = (char*)malloc( data_len * sizeof( char ) );
+
+    VectorArray<Ra::Core::Vector3>& vertices = geometry->getGeometry().verticesWithLock();
+    vertices.reserve( nb_data );
+
+    // checking for colors
+    bool colors       = false;
+    auto handle_color = geometry->getGeometry().vertexAttribs().addAttrib<Ra::Core::Vector4>(
```

Comment:
Check if colors exist in file before adding them as `vertexAttrib`

Reply:
Same for gps info

Reply:
if they are not in file, i delete them at the end (lines 220-227)

Reply:
This looks sub-optimal.
Can't you just move the attrib creation after the check ?

Reply:
in that case I would need a temporary vector to store colors and gps_time in the data reading loop, and that is what we tried to avoid in the first place

Reply:
I don't think so.
You can call `addAttrib` at line 137, ie. when setting colors to true.
In fact you don't need the `colors` variable: you can only check if `handle_color` is valid, using 
    handle_color.idx().isValid()

See https://storm-irit.github.io/Radium-Engine/release-candidate/classRa_1_1Core_1_1Utils_1_1AttribHandle.html#af86250020bb2831ce80a811c17d3a6c5
and https://storm-irit.github.io/Radium-Engine/release-candidate/Index_8hpp_source.html

Reply:
updated code to call findAttrib every iteration, please check if it's the suitable behaviour (i did not use idx()...)